R/board_ms365.R

```diff
@@ -0,0 +1,183 @@
+#' Use a OneDrive or Sharepoint document library as a board
+#'
+#' Pin data to a folder in Onedrive or a SharePoint Online document library
+#' using the Microsoft365R package.
+#'
+#' @inheritParams new_board
+#' @param drive A OneDrive or SharePoint document library object, as obtained from Microsoft365. See the examples.
+#' @param path Path to directory to store pins. Will be created if it
+#'   doesn't already exist.
+#' @export
+#' @examples
+#' \dontrun{
+#' # a board in your personal OneDrive
+#' od <- get_personal_onedrive()
+#' odboard <- board_onedrive(od, "myboard")
+#' odboard %>% pin_write(iris)
+#'
+#' # a board in OneDrive for Business
+#' odb <- get_business_onedrive(tenant="mytenant")
+#' odbboard <- board_onedrive(odb, "myproject/board")
+#'
+#' # a board in a SharePoint Online document library
+#' sp <- get_sharepoint_site("my site", tenant="mytenant")
+#' doclib <- sp$get_drive()
+#' spboard <- board_sharepoint(doclib, "general/project1/board")
+#' }
+board_ms365 <- function(drive, path, versioned = TRUE, cache = NULL) {
+  check_installed("Microsoft365R")
+
+  if (!inherits(drive, "ms_drive")) {
+    abort("`drive` must be a OneDrive or SharePoint document library object")
+  }
+
+  # try to create the board folder: ignore error if folder already exists
+  try(drive$create_folder(path), silent = TRUE)
+  folder <- drive$get_item(path)
+  if (!folder$is_folder()) {
+    abort("Invalid path specified")
+  }
+
+  cache <- cache %||% board_cache_path(paste0("ms365-", hash(folder$properties$id)))
+  new_board_v1("pins_board_ms365",
+    folder = folder,
+    # path = path,
+    cache = cache,
+    versioned = versioned
+  )
+}
+
+#' @export
+#' @rdname board_ms365
+board_sharepoint <- board_ms365
+
+#' @export
+#' @rdname board_ms365
+board_onedrive <- board_ms365
+
+board_ms365_test <- function(...) {
+  skip_if_missing_envvars("board_ms365()", "PINS_MS365_TEST_DRIVE")
+
+  drv <- readRDS(Sys.getenv("PINS_MS365_TEST_DRIVE"))
+  board_ms365(drv, path = "pin_testing", cache = tempfile(), ...)
+}
+
+#' @export
+pin_list.pins_board_ms365 <- function(board, ...) {
+  ms365_list_dirs(board)
+}
+
+#' @export
+pin_exists.pins_board_ms365 <- function(board, name, ...) {
+  name %in% ms365_list_dirs(board)
+}
+
+#' @export
+pin_delete.pins_board_ms365 <- function(board, names, by_item = FALSE, ...) {
```

Comment:
Is it possible to catch the error that happens when document protection is on and provide a better error message? (I'd only consider doing this if `ms365_delete_dir()` throws a classed error; string matching on error text is too fragile.)

Reply:
The error message I've seen isn't too bad; it actually tells you that you can't delete a folder that contains files. But I'm not familiar with all the possible error messages that MS 365 can produce.

The other possibility is to switch the default to `by_item=TRUE`, which should work everywhere. It'll make deleting folders that contain lots of files rather slow, but maybe that's not a problem here.

Reply:
I've revamped it so that by-item deletion is now an attribute of the board itself. This means you no longer have to pass an argument to `pin_delete` and `pin_version_delete`, which should be a much cleaner solution all-round.

Reply:
Perfect!